### PR TITLE
Change default AMG settings

### PR DESCRIPTION
**Pull request by MarcelKoch**
Based on the discussion in #1264 this PR changes some of the default settings of the AMG. The overall goal it to make it possible that our default AMG (no extra settings except the coarsening and stopping criterion) helps to reduce the necessary number of iterations adequately.

I've made each change in the defaults its own commit, with a brief description of my reasoning.

TODO:
- [x] update the documentation of the example
- [x] evaluate if we should also default the coarsening and/or stopping criterion
  - no to both of them.

### CODE REVIEW

**Comment by MarcelKoch** [comment]
There seem to be some incompatibilities between the `initial_guess_mode` and other settings of the multigrid. I'm not sure exactly which settings that touches on. But I've experienced that the mode `provided` can lead to no convergence at all. I think this needs to be investigated and then we need to warn the user if some incompatible settings are used.

**Comment by MarcelKoch** [comment]
While working on this, I come to realize that the default parameters are much better realized with the factory improvements mentioned in #1162. For example using
```
using create_factory_t = std::function<std::shared_ptr<LinOpFactory>(std::shared_ptr<const Executor>)>;
std::vector<create_factory_t> pre_smoother;
```
would allow defining the defaults directly in the parameter declaration:
```
std::vector<create_factory_t> pre_smoother{[](std::shared_ptr<const Executor> exec) {
  return Ir::build().with_solver(Jacobi::build().on(exec)).on(exec);
};
```
Perhaps with some helper functionality/change in factory it could be really concise.

**Review by fritzgoebel — CHANGES_REQUESTED** [review]

File: include/ginkgo/core/solver/multigrid.hpp [hunk 1]
```diff
@@ -323,6 +324,9 @@ class Multigrid : public EnableLinOp<Multigrid>,
 
         /**
          * Coarsest factory list.
+         *
+         * If not set, then GMRES preconditioned with a scalar Jacobi will be
+         * used as solver on the coarsest level.
```
[the remark this answers is not in the document]
Comment (MarcelKoch) [reply include/ginkgo/core/solver/multigrid.hpp:333]: No, I forgot that this is actually working now.

File: core/solver/multigrid.cpp [hunk 2]
```diff
@@ -554,17 +568,26 @@ void Multigrid::generate()
             using value_type =
                 typename std::decay_t<decltype(*mg_level)>::value_type;
             auto exec = this->get_executor();
+            // default coarse grid solver, direct LU
+            // TODO: maybe remove fixed index type
+            auto gen_default_solver = [&] {
+                return experimental::solver::Direct<value_type, int32>::build()
```
Comment (yhmtsai) [inline core/solver/multigrid.cpp:574]: Could we use the experimental class in another public class?

File: examples/multigrid-preconditioned-solver-customized/multigrid-preconditioned-solver-customized.cpp [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+    auto ic_gen = gko::share(
+        ic::build()
+            .with_factorization_factory(
+                gko::factorization::Ic<ValueType, int>::build().on(exec))
+            .on(exec));
+    auto smoother_gen = gko::share(
+        gko::solver::build_smoother(ic_gen, 2u, static_cast<ValueType>(0.9)));
+    // Use Pgm as the MultigridLevel factory (this is also the default).
+    auto mg_level_gen =
+        gko::share(pgm::build().with_deterministic(true).on(exec));
```
Comment (yhmtsai) [inline examples/multigrid-preconditioned-solver-customized/multigrid-preconditioned-solver-customized.cpp:141]: Is it default mg_level? I do not see when it is set

File: core/solver/multigrid.cpp (hunk 2 above)
Comment (MarcelKoch) [reply core/solver/multigrid.cpp:574]: I don't see why not. The default initialization is internal anyway.

File: examples/multigrid-preconditioned-solver-customized/multigrid-preconditioned-solver-customized.cpp (hunk 3 above)
Comment (MarcelKoch) [reply examples/multigrid-preconditioned-solver-customized/multigrid-preconditioned-solver-customized.cpp:141]: You're right, I will remove that comment

File: examples/ir-ilu-preconditioned-solver/doc/builds-on [hunk 4]
```diff
@@ -1 +1 @@
-ilu-preconditioned-solver iterative-refinement
+ic-preconditioned-solver iterative-refinement
```
[the remark this answers is not in the document]
Comment (MarcelKoch) [reply examples/ir-ilu-preconditioned-solver/doc/builds-on:1]: don't know how that got in there...

File: core/solver/multigrid.cpp (hunk 2 above)
Comment (MarcelKoch) [reply core/solver/multigrid.cpp:574]: The experimental features can't be turned off. Only the experimental MPI stuff is disabled if MPI is not available.

**Review by yhmtsai — APPROVED** [review]

**Comment by MarcelKoch** [comment]
rebase!

**Pull request merged**